File: 3751-total-waviness-of-numbers-in-range-i/3751-total-waviness-of-numbers-in-range-i.java
import java.util.ArrayList;
import java.util.List;

class Solution {
    public int totalWaviness(int num1, int num2) {
        int totalSum = 0;
        
        // Loop through each number in the inclusive range [num1, num2]
        for (int i = num1; i <= num2; i++) {
            totalSum += calculateWaviness(i);
        }
        
        return totalSum;
    }

    private int calculateWaviness(int n) {
        // Numbers with fewer than 3 digits have a waviness of 0
        if (n < 100) {
            return 0;
        }

        // Extract digits and store them in an indexed list
        List<Integer> digits = new ArrayList<>();
        while (n > 0) {
            digits.add(n % 10);
            n /= 10;
        }

        int wavinessCount = 0;
        int size = digits.size();

        // Note: The digits list is collected in reverse order, 
        // but the peak/valley relative relationship between neighbors remains identical.
        for (int i = 1; i < size - 1; i++) {
            int current = digits.get(i);
            int prev = digits.get(i - 1);
            int next = digits.get(i + 1);

            // Check if current digit is a Peak
            if (current > prev && current > next) {
                wavinessCount++;
            }
            // Check if current digit is a Valley
            else if (current < prev && current < next) {
                wavinessCount++;
            }
        }

        return wavinessCount;
    }
}

File: 3751-total-waviness-of-numbers-in-range-i/3751-total-waviness-of-numbers-in-range-i.js
/**
 * @param {number} num1
 * @param {number} num2
 * @return {number}
 */
var totalWaviness = function(num1, num2) {
    // Helper function to calculate the waviness of a single number
    function getWaviness(x) {
        const digits = [];
        while (x > 0) {
            digits.push(x % 10);
            x = Math.floor(x / 10);
        }
        
        // Since we extracted digits via modulo, they are reversed.
        // Reversing them back places them in their original reading order.
        digits.reverse();
        
        let waviness = 0;
        const len = digits.length;
        
        // Numbers with fewer than 3 digits cannot have peaks or valleys
        if (len < 3) return 0;
        
        // Check every internal digit to see if it is a peak or a valley
        for (let i = 1; i < len - 1; i++) {
            const current = digits[i];
            const prev = digits[i - 1];
            const next = digits[i + 1];
            
            // Peak condition
            if (current > prev && current > next) {
                waviness++;
            } 
            // Valley condition
            else if (current < prev && current < next) {
                waviness++;
            }
        }
        
        return waviness;
    }

    let total = 0;
    // Iterate through every number in the inclusive range
    for (let i = num1; i <= num2; i++) {
        total += getWaviness(i);
    }
    
    return total;
};
